rewrite-maven/src/main/java/org/openrewrite/maven/tree/ResolvedGroupArtifactVersion.java

```diff
@@ -16,16 +16,26 @@
 
 package org.openrewrite.maven.tree;
 
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.google.errorprone.annotations.InlineMe;
 import lombok.Value;
-import lombok.With;
 import org.jspecify.annotations.Nullable;
 
 import java.io.Serializable;
+import java.lang.ref.WeakReference;
+import java.util.LinkedHashMap;
+import java.util.Map;
 import java.util.Objects;
 
 @Value
-@With
 public class ResolvedGroupArtifactVersion implements Serializable {
+    private static final LinkedHashMap<String, WeakReference<ResolvedGroupArtifactVersion>> CACHE = new LinkedHashMap<String, WeakReference<ResolvedGroupArtifactVersion>>() {
```

Comment:
Since we have Caffeine anyway, I would just use that, as it will perform better. I would probably also skip the weak references as the objects aren't that heavy (especially with string deduplication by the JVM).

Reply:
Most String usage tends to land in the JVM string pool which automatically handles deduplication. Essentially, `"value"` lands in the string pool and `new String(..)` lands on heap. The later aren't automatically deduped, but they can also be converted to interned strings.

Frameworks, such as Jackson, put a lot of care into making sure to take advantage of this fact in both directions.